Implement BN_CTX_get

julek-wolfssl opened the pull request.

julek-wolfssl committed:
Implement BN_CTX_get

SparkiDev commented on src/ssl_bn.c:
```diff
@@ -2366,61 +2366,84 @@ int wolfSSL_BN_print_fp(XFILE fp, const WOLFSSL_BIGNUM *bn)
-    (void)ctx;
     WOLFSSL_ENTER("wolfSSL_BN_CTX_init");
+    if (ctx != NULL) {
+        XMEMSET(ctx, 0, sizeof(WOLFSSL_BN_CTX));
```
Old implementations in OpenSSL reset the pool and isn't meant to be called in BN_CTX_new().
API only in 1.0.2 and earlier.

julek-wolfssl replied:
Its called in the original implementations of `BN_CTX_new` (https://github.com/openssl/openssl/commit/9b141126d4b6f0636bc047e81b846c193ae26611#diff-9ef8217a3064ad4a970c33b34c9c8cabc4d8452ae282e5c1a19cbb0d8793fd88R74). Would you like me to separate the functions out? Maybe we should deprecate `wolfSSL_BN_CTX_init` as its equivalent is no longer in OpenSSL for a long time.

SparkiDev replied:
*sigh* our implementation is not doing the right thing then.
BN_CTX_init is an OpenSSL API - our implementation has to match theirs.
Definitely deprecate it!

julek-wolfssl commented on src/ssl_bn.c:
```diff
@@ -2366,61 +2366,84 @@ int wolfSSL_BN_print_fp(XFILE fp, const WOLFSSL_BIGNUM *bn)
 
     WOLFSSL_ENTER("wolfSSL_BN_CTX_get");
+    if (ctx != NULL) {
+        struct WOLFSSL_BN_CTX_LIST** prev = &ctx->list;
```
I don't see how OpenSSL is limiting the size. `BN_CTX_POOL_SIZE` is just how many are in a single pool. If they run out of BN's in a pool they allocate a new one. That is true for 1.1.1k.

SparkiDev replied:
OK.

julek-wolfssl commented on src/ssl_bn.c:
```diff
@@ -2366,61 +2366,84 @@ int wolfSSL_BN_print_fp(XFILE fp, const WOLFSSL_BIGNUM *bn)
 {
-    /* wolfcrypt doesn't need BN context. */
-    static int ctx;
+    WOLFSSL_BN_CTX* ctx = NULL;
```
I don't understand. You want a macro to keep the old solution?

SparkiDev replied:
BN_CTX is not needed most of the time - only when using the BN_CTX_get() API.
Most projects don't use this API.
Have a ifdef to return to the behaviour that uses less resources.

julek-wolfssl replied:
Added option to compile it out.

SparkiDev replied:
Not seeing code.

julek-wolfssl replied:
Modifying to cover entire `BN_CTX` section not just `BN_CTX_get`.

julek-wolfssl committed:
wolfSSL_BN_CTX_get: prepend to list skipping need to traverse the list

julek-wolfssl committed:
Remove BN_CTX_init as its no longer in OpenSSL for a long time

SparkiDev reviewed (changes requested):
What's happening with BN_CTX_start and BN_CTX_end?

julek-wolfssl commented:
We decided that `BN_CTX_start` and `BN_CTX_end` are not worth implementing in wolfSSL. They add unnecessary complexity due and don't add any benefit in wolfSSL since we don't use `BN_CTX` internally.

SparkiDev reviewed (changes requested):
Still need to have wolfSSL_BN_CTX_new().
Just don't allocate.

julek-wolfssl committed:
fixup! Implement BN_CTX_get

julek-wolfssl committed:
Remove internal use of wolfSSL_BN_CTX_new()

julek-wolfssl commented:
Added definitions for `BN_CTX_new` and `BN_CTX_free` when `NO_WOLFSSL_BN_CTX` is defined. Removed internal usage of `wolfSSL_BN_CTX_new` since it is not used.
